### DISCUSSION TITLE
Outstanding issue with upgrading from 1.x to 2.3

### DISCUSSION THREAD

**Discussion by ghost** [elsa-workflows/elsa-core#2422] [Q&A]
Hi,

Firstly I should say that I am not a *great* developer - it's not my day job.

I *had* a great developer and sadly he succumed to COVID a hew months ago, so I am now trying to carry on without him.

I've tried to get additional help but the market is a little over-heated in the UK so that's a bit challenging, so in the meantime I've decided to bite the bullet and try to do the re-implmentation into our app myself.

So far, I've upgraded the app to .net 5 and the libraries are now v2.3 together with all other non-elsa libraries.  I now have a few outstanding issues as follows:

1. in the server code, startup.cs has v 1.x code for creating the mongo store, 
           
 services.AddElsa(elsa =>
            {
                IMongoOption currentOptions = Configuration.GetSection("mongo").Get<MongoOption>();
                elsa.AddMongoDbStores(Configuration, databaseName: currentOptions.DatabaseName, connectionStringName: "Mongo");
            })

which gets me an error 'ElsaOptionsBuilder' does not contain a definition for 'AddMongoDbStores' and no accessible extension method 'AddMongoDbStores' accepting a first argument of type 'ElsaOptionsBuilder' could be found (are you missing a using directive or an assembly reference?)	

Can someone tell me what the new syntax should be please?

2.  the second problem, also in the servier is with a number of controllers, and they all are pretty much the same. again it's down to revised syntax since v1.x

        [HttpGet("ExecuteWorkflow")]
        [AllowAnonymous]
        public async Task ExecuteWorkflow([FromQuery] string definitionId, CancellationToken cancellationToken)
        {
            WorkflowDefinitionVersion workflowDefinitionVersion = await _workflowRegistry.GetWorkflowDefinitionAsync(definitionId, VersionOptions.Published, cancellationToken);
            if (workflowDefinitionVersion != null)
            {
                WorkflowExecutionContext workflowExecutionContext = await _workflowRunner.StartAsync(workflowDefinitionVersion);
                string s = string.Empty;
            }
        }

Again, I'd be really grateful if someone can point me in the right direction. Apologies for asking potentially dumb questions :)

**Comment by sfmskywalker** [comment]
Hi @JonathanR99 ,

1. To use MongoDB as the provider in Elsa 2, try the following code:

```csharp
services
    .AddElsa(elsa => elsa
        .UseMongoDbPersistence(mongo => Configuration.GetSection("mongo").Bind(mongo));
```

For that binding to work, make sure that your appsettings configuration has this structure:

```json
{
   "mongo": {
      "ConnectionString": "valid mongo connection string",
      "DatabaseName": "optional: database name if not provided as part of connection string"
   }
}
```

2. I'm not sure I understand what the second problem is? Are you getting an error, something doesn't compile?

**Reply by ghost** [reply]
The mongo connection issue seems to be resolved - thanks.

As far as the other issue is concerned, here's a screen grab from VS.

![ElsaIssue](https://user-images.githubusercontent.com/84724222/139819422-c4e1f136-f4fc-4d9d-8291-59f34490be19.PNG)

You are correct it won't build.

Is there some documentation avilable yet so I don't have to ask these questions?

regards


Jon

**Reply by sfmskywalker** [reply]
Try using the `GetAsync` method of the workflow registry (assuming it's of type `IWorkflowRegistry`).
Also make sure that you inject `IStartsWorkflow` (not sure what type the `_workflowRunner` is, but in your case `IStartsWorkflow` is what you want).

I refactored your code snippet (haven't tested it so might not quite work 100% - let me know if that's the case and the fix is not obvious):

```csharp
[HttpGet("ExecuteWorkflow")]
[AllowAnonymous]
public async Task ExecuteWorkflow([FromQuery] string definitionId, CancellationToken cancellationToken)
{
    var workflowBlueprint = await _workflowRegistry.GetWorkflowDefinitionAsync(definitionId, default, VersionOptions.Published, cancellationToken);
    if (workflowBlueprint != null)
    {
        var result = await _startsWorkflow.StartWorkflowAsync(workflowBlueprint);
        string s = string.Empty;
    }
}
```

A start has been made to document [changes between v1 and v2](https://elsa-workflows.github.io/elsa-core/docs/next/upgrading/upgrading-v1.x-to-v2.0), but it's far from complete. Your best bet is to look at code samples in the repo, the source code itself and of course reaching out here.
